storagenode/pieces/store.go

```diff
@@ -273,6 +274,83 @@ func (store *Store) Delete(ctx context.Context, satellite storj.NodeID, pieceID
 	return Error.Wrap(err)
 }
 
+// DeleteWithStorageFormat deletes the specified piece only at the specified format version
+func (store *Store) DeleteWithStorageFormat(ctx context.Context, satellite storj.NodeID, pieceID storj.PieceID, formatVer storage.FormatVersion) (err error) {
+	defer mon.Task()(&ctx)(&err)
+	err = store.blobs.DeleteWithStorageFormat(ctx, storage.BlobRef{
+		Namespace: satellite.Bytes(),
+		Key:       pieceID.Bytes(),
+	}, formatVer)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	// delete records in both the piece_expirations and pieceinfo DBs, wherever we find it.
+	// both of these calls should return no error if the requested record is not found.
+	if store.expirationInfo != nil {
+		_, err = store.expirationInfo.DeleteExpiration(ctx, satellite, pieceID)
+	}
+	if store.v0PieceInfo != nil {
+		err = errs.Combine(err, store.v0PieceInfo.Delete(ctx, satellite, pieceID))
+	}
+	return Error.Wrap(err)
+
+}
+
+// MigrateV0ToV1 will migrate a piece stored with storage format v0 to storage
+// format v1. If the piece is not stored as a v0 piece it will return an error.
+// The follow failures are possible:
+// - Fail to open or read v0 piece. In this case no artifacts remain.
+// - Fail to Write or Commit v1 piece. In this case no artifacts remain.
+// - Fail to Delete v0 piece. In this case v0 piece may remain, but v1 piece
+//   will exist and be preferred in future calls.
+func (store *Store) MigrateV0ToV1(ctx context.Context, satelliteID storj.NodeID, pieceID storj.PieceID) (err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	info, err := store.v0PieceInfo.Get(ctx, satelliteID, pieceID)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	r, err := store.Reader(ctx, satelliteID, pieceID)
+	if err != nil {
+		return Error.Wrap(err)
+	}
+
+	buf, err := ioutil.ReadAll(r)
```

Comment:
Isn't this going to read a lot of data into a giant buffer potentially? Is there a way to stream the reads from the reader into the writer instead of reading all the bytes?

Reply:
Good call. I changed it to use `io.Copy` instead.